src/contracts/access_control/GranularGuardianAccessControl.sol

```diff
@@ -0,0 +1,106 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity ^0.8.8;
+
+import {ICrossChainForwarder} from '../interfaces/ICrossChainForwarder.sol';
+import {ICrossChainControllerWithEmergencyMode} from '../interfaces/ICrossChainControllerWithEmergencyMode.sol';
+import {IGranularGuardianAccessControl, Envelope, ICrossChainReceiver} from './IGranularGuardianAccessControl.sol';
+import {AccessControlEnumerable} from 'openzeppelin-contracts/contracts/access/AccessControlEnumerable.sol';
+import {IWithGuardian} from 'solidity-utils/contracts/access-control/OwnableWithGuardian.sol';
+
+error DefaultAdminCantBe0();
+error CrossChainControllerCantBe0();
+error NewGuardianCantBe0();
+
+/**
+ * @title GranularGuardianAccessControl
+ * @author BGD Labs
+ * @notice Contract to manage a granular access to the methods safeguarded by guardian on CrossChainController
+ */
+contract GranularGuardianAccessControl is AccessControlEnumerable, IGranularGuardianAccessControl {
+  /// @inheritdoc IGranularGuardianAccessControl
+  address public immutable CROSS_CHAIN_CONTROLLER;
+
+  /// @inheritdoc IGranularGuardianAccessControl
+  bytes32 public constant SOLVE_EMERGENCY_ROLE = keccak256('SOLVE_EMERGENCY_ROLE');
+
+  /// @inheritdoc IGranularGuardianAccessControl
+  bytes32 public constant RETRY_ROLE = keccak256('RETRY_ROLE');
+
+  /**
+   * @param initialGuardians object with the initial guardians to assign the roles to
+   * @param crossChainController address of the CrossChainController
+   */
+  constructor(InitialGuardians memory initialGuardians, address crossChainController) {
+    if (crossChainController == address(0)) {
+      revert CrossChainControllerCantBe0();
+    }
+    if (initialGuardians.defaultAdmin == address(0)) {
+      revert DefaultAdminCantBe0();
+    }
+
+    CROSS_CHAIN_CONTROLLER = crossChainController;
+
+    _grantRole(DEFAULT_ADMIN_ROLE, initialGuardians.defaultAdmin);
+
+    if (initialGuardians.solveEmergencyGuardian != address(0)) {
+      _grantRole(SOLVE_EMERGENCY_ROLE, initialGuardians.solveEmergencyGuardian);
+    }
+    if (initialGuardians.solveEmergencyGuardian != address(0)) {
+      _grantRole(RETRY_ROLE, initialGuardians.retryGuardian);
+    }
+  }
+
+  /// @inheritdoc IGranularGuardianAccessControl
+  function retryEnvelope(
+    Envelope memory envelope,
```

Comment:
```suggestion
    Envelope calldata envelope,
```
Should be calldata here no?

Reply:
did some tests, and its cheaper to leave as memory. I think its because it needs to convert from calldata to memory, as its declared as memory in CCF. So for now i think it makes sense to just leave it as memory